scene/resources/curve.cpp

```diff
@@ -1496,76 +1499,135 @@ void Curve3D::_bake() const {
 		Vector3 npp = points[i + 1].position;
 		real_t d = position.distance_to(npp);
 
-		position = npp;
-		Plane post;
-		post.normal = position;
-		post.d = points[i + 1].tilt;
+		if (d > CMP_EPSILON) { // Avoid the degenerate case of two very close points.
+			position = npp;
+			Plane post;
+			post.normal = position;
+			post.d = points[i + 1].tilt;
 
-		dist += d;
+			dist += d;
 
-		pointlist.push_back(post);
-		distlist.push_back(dist);
+			pointlist.push_back(post);
+			distlist.push_back(dist);
+		}
 	}
 
 	baked_max_ofs = dist;
 
-	baked_point_cache.resize(pointlist.size());
-	Vector3 *w = baked_point_cache.ptrw();
-	int idx = 0;
+	const int point_count = pointlist.size();
+	{
+		baked_point_cache.resize(point_count);
+		Vector3 *w = baked_point_cache.ptrw();
 
-	baked_tilt_cache.resize(pointlist.size());
-	real_t *wt = baked_tilt_cache.ptrw();
+		baked_tilt_cache.resize(point_count);
+		real_t *wt = baked_tilt_cache.ptrw();
 
-	baked_up_vector_cache.resize(up_vector_enabled ? pointlist.size() : 0);
-	Vector3 *up_write = baked_up_vector_cache.ptrw();
+		baked_dist_cache.resize(point_count);
+		real_t *wd = baked_dist_cache.ptrw();
 
-	baked_dist_cache.resize(pointlist.size());
-	real_t *wd = baked_dist_cache.ptrw();
+		int idx = 0;
+		for (const Plane &E : pointlist) {
+			w[idx] = E.normal;
+			wt[idx] = E.d;
+			wd[idx] = distlist[idx];
 
-	Vector3 sideways;
-	Vector3 up;
-	Vector3 forward;
+			idx++;
+		}
+	}
+
+	if (!up_vector_enabled) {
+		baked_up_vector_cache.resize(0);
+		return;
+	}
 
-	Vector3 prev_sideways = Vector3(1, 0, 0);
-	Vector3 prev_up = Vector3(0, 1, 0);
-	Vector3 prev_forward = Vector3(0, 0, 1);
+	// Step 2: Calculate the up vectors and the whole local reference frame
+	//
+	// See C. Dougan, The Parallel Transport Frame, Game Programming Gems 2 for example
+	// for an example discussing about why not the Frenet frame.
+	{
+		PackedVector3Array forward_vectors;
 
-	for (const Plane &E : pointlist) {
-		w[idx] = E.normal;
-		wt[idx] = E.d;
-		wd[idx] = distlist[idx];
+		baked_up_vector_cache.resize(point_count);
+		forward_vectors.resize(point_count);
 
-		if (!up_vector_enabled) {
-			idx++;
-			continue;
+		Vector3 *up_write = baked_up_vector_cache.ptrw();
+		Vector3 *forward_write = forward_vectors.ptrw();
+
+		const Vector3 *points_ptr = baked_point_cache.ptr();
```

Comment:
Do not access the pointer when you do not need to write.

Reply:
you mean use `baked_point_cache[i]`? directly?

Reply:
Yes, it is not accessing the element directly in PackedArray(Vector), but like a sugar syntax of getter.

Reply:
I use the pointer for performance reason. 

Using baked_point_cache[i] will do index bound check every read, and in baking there are tend to be thousands of points per curve, a substantial cost.

```
	_FORCE_INLINE_ const T *ptr() const { return _cowdata.ptr(); }
	_FORCE_INLINE_ const T &operator[](int p_index) const { return _cowdata.get(p_index); }
```
```
	_FORCE_INLINE_ const T &get(int p_index) const {
		CRASH_BAD_INDEX(p_index, size()); // *** boundary check every time, bad for performance***

		return _ptr[p_index];
	}
```

Reply:
Okay, make sense.